subsys/settings/src/settings_psa.c

```diff
@@ -6,59 +6,68 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <psa/internal_trusted_storage.h>
-#include <zephyr/settings/settings.h>
 #include <zephyr/kernel.h>
 #include <zephyr/logging/log.h>
 #include <zephyr/psa/its_ids.h>
 
+#include "settings_psa_priv.h"
+
 LOG_MODULE_DECLARE(settings, CONFIG_SETTINGS_LOG_LEVEL);
 
 K_MUTEX_DEFINE(worker_mutex);
 static struct k_work_delayable worker;
 
-struct setting_entry {
-	char name[SETTINGS_MAX_NAME_LEN];
-	char value[SETTINGS_MAX_VAL_LEN];
-	size_t val_len;
-};
-
-static struct setting_entry entries[CONFIG_SETTINGS_TFM_ITS_NUM_ENTRIES];
 static int entries_count;
 
-static int settings_its_load(struct settings_store *cs, const struct settings_load_arg *arg);
-static int settings_its_save(struct settings_store *cs, const char *name, const char *value,
+static int settings_psa_load(struct settings_store *cs, const struct settings_load_arg *arg);
+static int settings_psa_save(struct settings_store *cs, const char *name, const char *value,
 			     size_t val_len);
 
-static const struct settings_store_itf settings_its_itf = {
-	.csi_load = settings_its_load,
-	.csi_save = settings_its_save,
+static const struct settings_store_itf settings_psa_itf = {
+	.csi_load = settings_psa_load,
+	.csi_save = settings_psa_save,
 };
 
-static struct settings_store default_settings_its = {.cs_itf = &settings_its_itf};
+static struct settings_store default_settings_psa = {.cs_itf = &settings_psa_itf};
+
+#if defined(CONFIG_SETTINGS_PSA_BACKEND_PS)
+#include <psa/protected_storage.h>
+#include <config_tfm.h>
+
+#define SETTINGS_PSA_CHUNK_SIZE PS_MAX_ASSET_SIZE
+#define SETTINGS_PSA_SET psa_ps_set
+#define SETTINGS_PSA_GET psa_ps_get
+#elif defined(CONFIG_SETTINGS_PSA_BACKEND_ITS)
+#include <psa/internal_trusted_storage.h>
+
+#define SETTINGS_PSA_CHUNK_SIZE CONFIG_TFM_ITS_MAX_ASSET_SIZE
+#define SETTINGS_PSA_SET psa_its_set
+#define SETTINGS_PSA_GET psa_its_get
+#else
+#error "No PSA backend selected"
+#endif
 
 /* Ensure Key configured max size does not exceed reserved Key range */
-BUILD_ASSERT(sizeof(entries) / CONFIG_TFM_ITS_MAX_ASSET_SIZE <=
-	     ZEPHYR_PSA_SETTINGS_TFM_ITS_UID_RANGE_SIZE,
-	     "entries array exceeds reserved ITS UID range");
+BUILD_ASSERT(sizeof(entries) / SETTINGS_PSA_CHUNK_SIZE <=
+	     ZEPHYR_PSA_SETTINGS_PSA_UID_RANGE_SIZE,
+	     "entries array exceeds reserved PSA storage UID range");
```

Comment:
Should I also add an assert for the case chunk number exceeds TF-M's storage entry count? That won't catch all misconfigurations, but will surely fail on ones where settings clearly won't fit even if no other ITS/PS entries are there

Reply:
A `BUILD_ASSERT` should do the trick. It currently only validates if the array fits in the designated ID range, and I want to make it `min(range_size, max_entries_from_tfm_conf)`